### ISSUE TITLE
expression push down storage bug 

### ISSUE THREAD

**Issue by wanpengfei-git** [StarRocks/starrocks#2726]
### Steps to reproduce the behavior (Required)
```
1. DROP DATABASE IF EXISTS d2;
CREATE DATABASE d2;
USE d2;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE IF NOT EXISTS t0 (c_0_0 TINYINT NULL ,c_0_1 DATE NOT NULL ,c_0_2 DECIMAL128(4, 0)  NULL ,c_0_3 BOOLEAN NULL ,c_0_4 DECIMAL(20, 2)  NULL ,c_0_5 DECIMAL128(12, 5)  NOT NULL ,c_0_6 TINYINT NOT NULL ,c_0_7 DATETIME NULL ,c_0_8 DECIMAL32(7, 7)  NULL ,c_0_9 DOUBLE NOT NULL ,c_0_10 DECIMAL(11, 2)  NULL ) DUPLICATE KEY (c_0_0,c_0_1,c_0_2) DISTRIBUTED BY HASH (c_0_1,c_0_0) properties("replication_num"="1") ;
CREATE TABLE t1 (c_1_0 DECIMAL128(37, 36)  NOT NULL ,c_1_1 VARCHAR(1) NOT NULL ,c_1_2 DECIMAL32(5, 0)  NULL ,c_1_3 VARCHAR(41) NULL ,c_1_4 CHAR(11) NOT NULL ,c_1_5 DATE NOT NULL ,c_1_6 VARCHAR(21) NOT NULL ,c_1_7 BOOLEAN NOT NULL ,c_1_8 TINYINT NOT NULL ,c_1_9 DATE NULL ,c_1_10 DATETIME NOT NULL ,c_1_11 VARCHAR(1) NOT NULL ) UNIQUE KEY (c_1_0,c_1_1) DISTRIBUTED BY HASH (c_1_1) properties("replication_num"="1") ;
INSERT INTO t1 VALUES (5E-36, "", 90661, "", "", "1970-01-01", "", false, 4, "1970-01-01", "1970-01-01 08:00:00", ""), (7E-36, "", 76895, "", "B", "1970-01-01", "", false, 1, NULL, "1970-01-01 08:00:00", ""), (6E-36, "", 79148, "", "A", "1970-01-01", "v", false, 1, "1970-01-01", "1970-01-01 08:00:00", "");
INSERT INTO t0 VALUES (6, "1970-01-01", 2248, false, NULL, 2.89249, 1, "1970-01-01 08:00:00", 0E-7, 0.1981966941182951, 8726286.54);
TRUNCATE TABLE t0;
DELETE FROM t0 WHERE c_0_1="1970-01-20";
INSERT INTO t1 VALUES (7E-36, "", NULL, "M", "Mf", "1970-01-01", "", true, 2, "1970-01-01", "1970-01-01 08:00:00", ""), (3E-36, "", NULL, "i22eS3", "r", "1970-01-01", "BctSY", false, 6, "1970-01-01", "1970-01-01 08:00:00", ""), (8E-36, "", 10279, "nB", "", "1970-01-01", "57sS4", true, 0, "1970-01-01", "1970-01-01 08:00:00", "");
INSERT INTO t0 VALUES (2, "1970-01-01", 386, false, 1091422668154616.97, 92.47634, 4, NULL, NULL, 0.034162436633962456, 4180034.03), (NULL, "1970-01-01", 5114, true, NULL, 2.80806, 5, "1970-01-01 08:00:00", 6E-7, 0.26070602713625357, 1048896.66), (0, "1970-01-01", 632, NULL, 2121310303700300.38, 46.48010, 3, "1970-01-01 08:00:00", 4E-7, 0.8028471525291432, 4145778.62);
-- SELECT DISTINCT subt1.c_1_0 FROM  (SELECT t1.c_1_0, t1.c_1_1, t1.c_1_2, t1.c_1_3, t1.c_1_4, t1.c_1_5, t1.c_1_6, t1.c_1_7, t1.c_1_8, t1.c_1_9, t1.c_1_10, t1.c_1_11 FROM t1) subt1 FULL JOIN (SELECT t0.c_0_0, t0.c_0_1, t0.c_0_2, t0.c_0_3, t0.c_0_4, t0.c_0_5, t0.c_0_6, t0.c_0_7, t0.c_0_8, t0.c_0_9, t0.c_0_10 FROM t0) subt0 ON subt1.c_1_3 = subt0.c_0_2 AND subt1.c_1_3 <= subt0.c_0_2;
-- cardinality: 6
-- SELECT DISTINCT * FROM (SELECT DISTINCT subt1.c_1_0 FROM  (SELECT t1.c_1_0, t1.c_1_1, t1.c_1_2, t1.c_1_3, t1.c_1_4, t1.c_1_5, t1.c_1_6, t1.c_1_7, t1.c_1_8, t1.c_1_9, t1.c_1_10, t1.c_1_11 FROM t1) subt1 FULL JOIN (SELECT t0.c_0_0, t0.c_0_1, t0.c_0_2, t0.c_0_3, t0.c_0_4, t0.c_0_5, t0.c_0_6, t0.c_0_7, t0.c_0_8, t0.c_0_9, t0.c_0_10 FROM t0) subt0 ON subt1.c_1_3 = subt0.c_0_2 AND subt1.c_1_3 <= subt0.c_0_2 WHERE (CASE  WHEN (subt0.c_0_2)  BETWEEN (subt0.c_0_2) AND (subt0.c_0_2)  THEN 0.1 ELSE CAST(73 AS DECIMAL32 )  END) IN (0.0)  UNION ALL SELECT DISTINCT subt1.c_1_0 FROM  (SELECT t1.c_1_0, t1.c_1_1, t1.c_1_2, t1.c_1_3, t1.c_1_4, t1.c_1_5, t1.c_1_6, t1.c_1_7, t1.c_1_8, t1.c_1_9, t1.c_1_10, t1.c_1_11 FROM t1) subt1 FULL JOIN (SELECT t0.c_0_0, t0.c_0_1, t0.c_0_2, t0.c_0_3, t0.c_0_4, t0.c_0_5, t0.c_0_6, t0.c_0_7, t0.c_0_8, t0.c_0_9, t0.c_0_10 FROM t0) subt0 ON subt1.c_1_3 = subt0.c_0_2 AND subt1.c_1_3 <= subt0.c_0_2 WHERE (NOT ((CASE  WHEN (subt0.c_0_2)  BETWEEN (subt0.c_0_2) AND (subt0.c_0_2)  THEN 0.1 ELSE CAST(73 AS DECIMAL32 )  END) IN (0.0) )) UNION ALL SELECT DISTINCT subt1.c_1_0 FROM  (SELECT t1.c_1_0, t1.c_1_1, t1.c_1_2, t1.c_1_3, t1.c_1_4, t1.c_1_5, t1.c_1_6, t1.c_1_7, t1.c_1_8, t1.c_1_9, t1.c_1_10, t1.c_1_11 FROM t1) subt1 FULL OUTER JOIN (SELECT t0.c_0_0, t0.c_0_1, t0.c_0_2, t0.c_0_3, t0.c_0_4, t0.c_0_5, t0.c_0_6, t0.c_0_7, t0.c_0_8, t0.c_0_9, t0.c_0_10 FROM t0) subt0 ON subt1.c_1_3 = subt0.c_0_2 AND subt1.c_1_3 <= subt0.c_0_2 WHERE (((CASE  WHEN (subt0.c_0_2)  BETWEEN (subt0.c_0_2) AND (subt0.c_0_2)  THEN 0.1 ELSE CAST(73 AS DECIMAL32 )  END) IN (0.0) ) IS NULL)) t;
-- cardinality: 1
```

```
DROP DATABASE IF EXISTS d1;
CREATE DATABASE d1;
USE d1;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE t0 (c_0_0 DECIMAL128(7, 0)  NULL ,c_0_1 DECIMAL(10, 6)  NOT NULL ,c_0_2 VARCHAR(1) NULL ,c_0_3 DECIMAL(14, 2)  NOT NULL ,c_0_4 DECIMAL128(10, 10)  NOT NULL ,c_0_5 DECIMAL32(4, 4)  REPLACE_IF_NOT_NULL NULL ) AGGREGATE KEY (c_0_0,c_0_1,c_0_2,c_0_3,c_0_4) DISTRIBUTED BY HASH (c_0_3,c_0_4,c_0_1,c_0_0,c_0_2) properties("replication_num"="1") ;
CREATE TABLE IF NOT EXISTS t1 (c_1_0 DECIMAL(22, 3)  NULL ,c_1_1 CHAR(11) NOT NULL ,c_1_2 BOOLEAN NULL ,c_1_3 DECIMAL64(14, 14)  NOT NULL ,c_1_4 CHAR(1) NULL ,c_1_5 DECIMAL(18, 3)  NULL ,c_1_6 DECIMAL32(8, 8)  NOT NULL ,c_1_7 DECIMAL(10, 3)  NULL ,c_1_8 FLOAT NULL ) UNIQUE KEY (c_1_0,c_1_1,c_1_2,c_1_3,c_1_4,c_1_5,c_1_6) DISTRIBUTED BY HASH (c_1_5,c_1_2,c_1_0,c_1_6) properties("replication_num"="1") ;
CREATE TABLE IF NOT EXISTS t2 (c_2_0 BOOLEAN NOT NULL ,c_2_1 INT NOT NULL ,c_2_2 DECIMAL(10, 7)  NULL ,c_2_3 DECIMAL64(7, 1)  NOT NULL ,c_2_4 DECIMAL128(4, 3)  NULL ,c_2_5 TINYINT NOT NULL ) DUPLICATE KEY (c_2_0,c_2_1,c_2_2,c_2_3) DISTRIBUTED BY HASH (c_2_0,c_2_3) properties("replication_num"="1") ;
CREATE TABLE t3 (c_3_0 CHAR(1) NULL ,c_3_1 DATETIME NOT NULL ,c_3_2 DECIMAL(26, 7)  NOT NULL ,c_3_3 SMALLINT NULL ,c_3_4 BIGINT NOT NULL ,c_3_5 DATE NULL ,c_3_6 BOOLEAN NOT NULL ,c_3_7 BIGINT NULL ,c_3_8 DECIMAL128(34, 18)  NOT NULL ,c_3_9 DECIMAL(6, 5)  NULL ,c_3_10 DECIMAL64(1, 0)  NULL ,c_3_11 DOUBLE NULL ,c_3_12 VARCHAR(1) NULL ,c_3_13 VARCHAR(11) NULL ,c_3_14 DECIMAL32(7, 1)  NULL ,c_3_15 DECIMAL64(1, 1)  NULL ,c_3_16 DECIMAL(16, 1)  NOT NULL ,c_3_17 BOOLEAN NOT NULL ,c_3_18 DECIMAL32(6, 1)  NULL ) UNIQUE KEY (c_3_0,c_3_1,c_3_2) DISTRIBUTED BY HASH (c_3_2,c_3_0) properties("replication_num"="1") ;
CREATE TABLE t4 (c_4_0 BOOLEAN NOT NULL ,c_4_1 DECIMAL64(9, 0)  NULL ,c_4_2 DATETIME NOT NULL ,c_4_3 BOOLEAN NULL ,c_4_4 DECIMAL(15, 1)  NOT NULL ,c_4_5 DECIMAL32(5, 2)  NOT NULL ,c_4_6 DATETIME NOT NULL ,c_4_7 CHAR(11) REPLACE NULL ,c_4_8 DECIMAL64(17, 3)  MAX NOT NULL ,c_4_9 DECIMAL32(4, 0)  MIN NULL ,c_4_10 DECIMAL64(2, 0)  REPLACE NULL ,c_4_11 BITMAP BITMAP_UNION NULL ,c_4_12 DECIMAL(10, 4)  REPLACE_IF_NOT_NULL NULL ,c_4_13 DECIMAL128(17, 8)  REPLACE NULL ,c_4_14 VARCHAR(21) MAX NOT NULL ,c_4_15 DOUBLE MAX NULL ,c_4_16 BOOLEAN REPLACE NOT NULL ,c_4_17 SMALLINT MAX NOT NULL ) AGGREGATE KEY (c_4_0,c_4_1,c_4_2,c_4_3,c_4_4,c_4_5,c_4_6) DISTRIBUTED BY HASH (c_4_5,c_4_2,c_4_3,c_4_4,c_4_6) properties("replication_num"="1") ;
INSERT INTO t0 VALUES (2229031, 0.004756, "", 1504125469.58, 7E-10, 0.0001), (3425867, 0.005557, "", 278832399.16, 8E-10, 0.0000);
INSERT INTO t4 VALUES (false, 497134223, "1970-01-01 08:00:00", false, 4430365016478.9, 0.79, "1970-01-01 08:00:00", "", 30646729882.552, 8923, 86, bitmap_hash(""), 27.2876, 2.11140848, "zXDyB", -6.24576568E8, false, 15);
INSERT INTO t1 VALUES (NULL, "S6StBdt", true, 5E-14, "", NULL, 7E-8, 3671.033, 0.08035117043126117), (2890371923781143.970, "b", true, 5E-14, "", 99991479975.513, 7E-8, 7623.803, NULL);
INSERT INTO t0 VALUES (5119511, 0.007310, NULL, 5434833771.97, 0E-10, 0.0002), (NULL, 0.005715, "", 6336344994.12, 2E-10, 0.0007), (90432, 0.006659, NULL, 9949942904.11, 7E-10, 0.0002), (4471265, 0.002756, "", 4875436558.60, 5E-10, 0.0003);
INSERT INTO t1 VALUES (6278186406321010.726, "nU3O", false, 0E-14, "", 470463285903.156, 3E-8, 5389.621, 9.26948436E8);
INSERT INTO t3 VALUES ("", "1970-01-01 08:00:00", 776411659156.6687408, 0, -204539929, "1970-01-01", true, NULL, 0.001774156754029822, 0.00003, 6, 0.515575539520538, "", "", 96955.1, NULL, 17233110722949.7, true, NULL), ("", "1970-01-01 08:00:00", 784638149197.3605242, 11, -431647979, "1970-01-01", false, 2004260609, 0.008243558313951271, 0.00004, 3, 0.15347458885621457, "", "i", NULL, 0.6, 75215095043696.5, false, 9847.4);
INSERT INTO t1 VALUES (4611131332171782.904, "Q", false, 3E-14, "", NULL, 7E-8, 4859.199, 0.12063990458508023), (8202771309419399.552, "", NULL, 5E-14, "", 652400898504.974, 1E-8, 4622.166, NULL), (7068573248041367.539, "MX", NULL, 8E-14, "", 29544295214.678, 7E-8, 444.721, 2.03266459E9), (2043643733332146.485, "v4Dm", true, 2E-14, "", 636370800457.334, 6E-8, 3003.229, 0.8200695861959052);
INSERT INTO t2 VALUES (true, 12, 0.0000304, 83592.3, 0.002, 0);
INSERT INTO t2 VALUES (false, 9, 0.0000149, 12429.6, 0.007, 4), (true, 31, 0.0000518, 72780.2, 0.008, 4), (true, 2, 0.0000604, 98233.4, 0.004, 7);
INSERT INTO t3 VALUES ("", "1970-01-01 08:00:00", 254212295172.5691771, 9, 1794972618, "1970-01-01", false, 1669471118, 0.007384394422958625, 0.00004, 2, 0.8080523515282253, "", NULL, 23534.0, 0.8, 27311036316252.1, true, 6776.8);
INSERT INTO t3 VALUES ("", "1970-01-01 08:00:00", 85861354855.7591095, 14, 1419245553, "1970-01-01", true, -886183547, 0.005137427322369807, 0.00006, 6, 0.3651141613996193, "", NULL, NULL, 0.1, 23711791981349.2, true, 1872.9);
DELETE FROM t3 WHERE c_3_0="";
TRUNCATE TABLE t0;
INSERT INTO t0 VALUES (NULL, 0.005560, "", 7493620219.45, 6E-10, 0.0004), (3256416, 0.002572, "", 1693927794.32, 0E-10, 0.0006);
-- SELECT DISTINCT t3.c_3_0 FROM t3;
-- cardinality: 0
-- SELECT DISTINCT * FROM (SELECT DISTINCT t3.c_3_0 FROM t3 WHERE CAST(t3.c_3_0 AS BOOLEAN )  UNION ALL SELECT DISTINCT t3.c_3_0 FROM t3 WHERE (NOT (CAST(t3.c_3_0 AS BOOLEAN ) )) UNION ALL SELECT DISTINCT t3.c_3_0 FROM t3 WHERE ((CAST(t3.c_3_0 AS BOOLEAN ) ) IS NULL)) t;
-- cardinality: 1
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - main Release
 - mysql> select current_version();
+---------------------------------+
| current_version()               |
+---------------------------------+
| SQLANCER_MASTER_RELEASE f24400d |
+---------------------------------+

**Comment by Youngwb** [comment]
Simplification of the above problem
```
mysql> explain SELECT  CASE WHEN null THEN 0.1 ELSE CAST(73 AS DECIMAL) END;
+----------------------------+
| Explain String             |
+----------------------------+
| PLAN FRAGMENT 0            |
|  OUTPUT EXPRS:1: case      |
|   PARTITION: UNPARTITIONED |
|                            |
|   RESULT SINK              |
|                            |
|   0:UNION                  |
|      constant exprs:       |
|          73                |
+----------------------------+
9 rows in set (3.22 sec)

mysql>  SELECT  CASE WHEN null THEN 0.1 ELSE CAST(73 AS DECIMAL) END;
+--------------------------------------------------------------+
| CASE WHEN NULL THEN 0.1 ELSE CAST(73 AS DECIMAL64(10,0)) END |
+--------------------------------------------------------------+
|                                                           73 |
+--------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> explain SELECT  CASE WHEN null THEN 0.1 ELSE CAST(73 AS DECIMAL32) END;
+-------------------------------------+
| Explain String                      |
+-------------------------------------+
| PLAN FRAGMENT 0                     |
|  OUTPUT EXPRS:1: case               |
|   PARTITION: UNPARTITIONED          |
|                                     |
|   RESULT SINK                       |
|                                     |
|   0:UNION                           |
|      constant exprs:                |
|          CAST(73 AS DECIMAL32(9,9)) |
+-------------------------------------+
9 rows in set (0.00 sec)

mysql>  SELECT  CASE WHEN null THEN 0.1 ELSE CAST(73 AS DECIMAL32) END;
+-------------------------------------------------------------+
| CASE WHEN NULL THEN 0.1 ELSE CAST(73 AS DECIMAL32(9,9)) END |
+-------------------------------------------------------------+
|                                                        NULL |
+-------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by satanson** [comment]
@Youngwb @wanpengfei-git
FE should report  error when executing `select cast(73 as DECIMAL32) `, because 73 casting to DECIMAL32 overflows.

**Comment by satanson** [comment]
decimal32 is invisible to users. its just used internally

**Comment by laotan332** [comment]
fix by #2755
